**Dataset Overview**

The Liberty Mutual Fire Peril Loss Cost dataset contains 105,450 commercial insurance records and 302 variables. In addition, there are 292 numerical variables and 10 categorical variables and 1 is an identifier. The response variable is target and this represents the fire-peril loss cost.

The dataset is representative of a typical commercial insurance dataset because there is extreme skewness, heavy tails and a large number of zeros. This makes it a strong candidate for GLMs (Tweedie/Gamma) and tree-based ensemble models.

In [ ]:
# ======================================================
# CIND820 - Keystone Project
# Data Analysis and Visualization: Fire Peril Loss Cost (Liberty Mutual Insurance)
# Author: Debra Allen 
# ======================================================

# Call Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [ ]:
# ======================================================
# Descriptive Statistics
# ======================================================

DATA_URL = "https://media.githubusercontent.com/media/uni-Deb7943/TMU-CIND820/refs/heads/main/liberty_train.csv?token=CATLW7R3ADRF5LKRZORC5MTJYWLCI"
data = pd.read_csv(DATA_URL, low_memory=False)

# Display the shape of the dataset
print("Dataset Shape:", data.shape)

# Display the first few rows of the dataset
print("\nFirst few rows of the dataset:")
data.head()

Dataset Shape: (452061, 302)

First few rows of the dataset:


,id,target,var1,var2,var3,var4,var5,var6,var7,var8,...,weatherVar227,weatherVar228,weatherVar229,weatherVar230,weatherVar231,weatherVar232,weatherVar233,weatherVar234,weatherVar235,weatherVar236
0,1,0.0,Z,Z,Z,N1,Z,Z,3,1,...,0.170351,0.0,0.00000,0.000000,1.117353,1.215303,0.112556,2.355737,0.404655,0.138667
1,2,0.0,Z,Z,Z,C1,Z,Z,3,2,...,0.266173,0.0,27.50823,0.000000,3.828979,1.036739,0.033052,0.856632,0.231232,0.742199
2,3,0.0,3,Z,4,J3,B,B,2,4,...,0.979517,0.0,0.00000,0.456134,0.098790,1.076535,0.566352,0.696013,0.693695,0.070654
3,4,0.0,3,Z,4,H1,B,Z,3,4,...,0.308761,0.0,0.00000,4.349042,0.401975,0.340631,0.290147,0.000000,0.000000,0.090332
4,5,0.0,Z,Z,Z,H1,Z,Z,2,4,...,2.171972,0.0,0.00000,0.514990,0.516095,1.016120,1.313732,1.338487,2.948202,0.816485


In [21]:
# Reduce the size of the Train dataset to 25 % as the memory is being fully exhausted
# The rows are selected randomly

train = data.sample(frac = .25, replace = False)

In [ ]:
# Display basic structure of the dataset
print("\nDataset Info:")
train.info()

# Display the shape of the reduced dataset
print("\nDataset Shape:", train.shape)


Dataset Info:
<class 'pandas.DataFrame'>
Index: 113015 entries, 313446 to 305889
Columns: 302 entries, id to weatherVar236
dtypes: float64(291), int64(1), str(10)
memory usage: 261.3 MB

Dataset Shape: (113015, 302)


In [28]:
# Display summary statistics of the dataset
print("\nSummary Statistics:")
train.describe()


Summary Statistics:


,id,target,var10,var11,var12,var13,var14,var15,var16,var17,...,weatherVar227,weatherVar228,weatherVar229,weatherVar230,weatherVar231,weatherVar232,weatherVar233,weatherVar234,weatherVar235,weatherVar236
count,113015.000000,113015.000000,113015.000000,113015.000000,24285.000000,113015.000000,40410.000000,88368.000000,22634.000000,113015.000000,...,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000,108218.000000
mean,452059.193868,0.006180,4.128454,1316.001939,3.098764,1.304795,0.222034,31.163328,73.254705,68.561699,...,0.948338,0.792867,1.012005,0.970197,1.008444,1.012088,0.895103,1.072779,0.993245,1.045110
std,260785.668579,0.195755,0.311165,1084.700057,0.290032,0.814208,0.427496,25.511297,43.384470,37.321490,...,0.733207,26.708848,3.807272,1.890621,3.381126,0.183468,1.108364,0.908525,1.373227,4.688720
min,2.000000,0.000000,3.742477,500.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.010647,0.000000,0.000000,0.000000,0.000000,0.000000,0.001047,0.000000,0.000000,0.000000
25%,227064.000000,0.000000,3.898949,681.909085,2.936187,0.832555,0.000000,13.000000,8.000000,49.721090,...,0.415230,0.000000,0.000000,0.000000,0.069835,0.906219,0.246716,0.374776,0.231232,0.030111
50%,452745.000000,0.000000,4.039930,939.414711,3.096498,1.394959,0.000000,26.000000,100.000000,85.921889,...,0.777225,0.000000,0.000000,0.344728,0.200987,1.060452,0.746745,0.856632,0.578079,0.120323
75%,678003.500000,0.000000,4.314082,1500.000000,3.273289,1.939381,0.000000,41.000000,100.000000,100.000000,...,1.288277,0.000000,0.000000,1.366301,0.676203,1.134062,1.230880,1.552645,1.213965,0.397512
max,902755.000000,18.985234,5.677710,14104.377710,3.961367,2.894981,2.168059,440.500000,100.000000,100.000000,...,3.641247,3778.103919,171.926438,55.553810,137.327164,1.546449,50.053757,5.353948,8.093103,100.087491


In [29]:
# Display the data types of the columns in the dataset
print("\nData Types of Columns:")
train.dtypes


Data Types of Columns:


id                 int64
target           float64
var1                 str
var2                 str
var3                 str
                  ...   
weatherVar232    float64
weatherVar233    float64
weatherVar234    float64
weatherVar235    float64
weatherVar236    float64
Length: 302, dtype: object

In [42]:
# compute mean, median, max, and zero-rate of the target variable only
if 'target' in train.columns:
    print('Mean Loss Cost - Reduced Dataset:', train['target'].mean())
    print('Median - Reduced Dataset:', train['target'].median())
    print('Max - Reduced Dataset:', train['target'].max())
    perc_zero = (train['target'] == 0).mean() * 100
    print('Zero-rate - Reduced Dataset: {:.4f}%'.format(perc_zero))
else:
    print('train target column not found')

if 'target' in data.columns:
    print('\nMean Loss Cost - Full Dataset:', data['target'].mean())
    print('Median - Full Dataset:', data['target'].median())
    print('Max - Full Dataset:', data['target'].max())
    perc_zero = (data['target'] == 0).mean() * 100
    print('Zero-rate - Full Dataset: {:.4f}%'.format(perc_zero))
else:
    print('data target column not found')

Mean Loss Cost - Reduced Dataset: 0.006179510060363668
Median - Reduced Dataset: 0.0
Max - Reduced Dataset: 18.98523446
Zero-rate - Reduced Dataset: 99.7540%

Mean Loss Cost - Full Dataset: 0.007234069647007815
Median - Full Dataset: 0.0
Max - Full Dataset: 25.92013792
Zero-rate - Full Dataset: 99.7372%


In [44]:
# Separate Variable types
categorical_cols = train.select_dtypes(include=['object']).columns
numerical_cols = train.select_dtypes(include=['number']).columns

print("\nCategorical Columns:", categorical_cols)
print("Numerical Columns:", numerical_cols)

# Display the number of categorical and numerical columns
print("\nNumber of Categorical Columns:", len(categorical_cols))
print("Number of Numerical Columns:", len(numerical_cols))


Categorical Columns: Index(['var1', 'var2', 'var3', 'var4', 'var5', 'var6', 'var7', 'var8', 'var9',
       'dummy'],
      dtype='str')
Numerical Columns: Index(['id', 'target', 'var10', 'var11', 'var12', 'var13', 'var14', 'var15',
       'var16', 'var17',
       ...
       'weatherVar227', 'weatherVar228', 'weatherVar229', 'weatherVar230',
       'weatherVar231', 'weatherVar232', 'weatherVar233', 'weatherVar234',
       'weatherVar235', 'weatherVar236'],
      dtype='str', length=292)

Number of Categorical Columns: 10
Number of Numerical Columns: 292


C:\Users\uni_f\AppData\Local\Temp\ipykernel_35444\4156532869.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = train.select_dtypes(include=['object']).columns


In [50]:
# Display summary statistics for numerical columns
describe_numerical = train[numerical_cols].describe().T
print("\nSummary Statistics for Numerical Columns:")
print(describe_numerical)

# Determine skewness and kurtosis for the numerical columns
skewness = train[numerical_cols].skew()
kurtosis = train[numerical_cols].kurtosis()

print("\nSkewness for Numerical Columns:")
print(skewness)

print("\nKurtosis for Numerical Columns:")
print(kurtosis)


Summary Statistics for Numerical Columns:
                  count           mean            std         min  \
id             113015.0  452059.193868  260785.668579    2.000000   
target         113015.0       0.006180       0.195755    0.000000   
var10          113015.0       4.128454       0.311165    3.742477   
var11          113015.0    1316.001939    1084.700057  500.000000   
var12           24285.0       3.098764       0.290032    0.000000   
...                 ...            ...            ...         ...   
weatherVar232  108218.0       1.012088       0.183468    0.000000   
weatherVar233  108218.0       0.895103       1.108364    0.001047   
weatherVar234  108218.0       1.072779       0.908525    0.000000   
weatherVar235  108218.0       0.993245       1.373227    0.000000   
weatherVar236  108218.0       1.045110       4.688720    0.000000   

                         25%            50%            75%            max  
id             227064.000000  452745.000000  678003.

Mean loss cost: approx 0.007
Median: 0
Maximum observed loss cost: 25.9
